flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/FunctionIdentifier.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.functions;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.catalog.ObjectIdentifier;
+import org.apache.flink.util.StringUtils;
+
+import java.io.Serializable;
+import java.util.Objects;
+import java.util.Optional;
+
+import static org.apache.flink.table.utils.EncodingUtils.escapeIdentifier;
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Identifies a system function with function name or a catalog function with a fully qualified identifier.
+ * Function catalog is responsible for resolving an identifier to a function.
+ */
+@PublicEvolving
+public final class FunctionIdentifier implements Serializable {
+
+	private final ObjectIdentifier objectIdentifier;
+
+	private final String functionName;
+
+	public static FunctionIdentifier of(ObjectIdentifier oi){
+		return new FunctionIdentifier(oi);
+	}
+
+	public static FunctionIdentifier of(String functionName){
+		return new FunctionIdentifier(functionName);
```

Comment:
Do we need to normalize functionName?

Reply:
I put the normalization logic in the constructors

Reply:
Got it.